subiquity/server/controllers/filesystem.py

```diff
@@ -205,48 +184,65 @@ def guided_lvm(self, target, mode=None, lvm_options=None):
+        return gaps.at_offset(disk, target.gap.offset)
+
+    @start_guided.register
+    def start_guided_resize(self, target: GuidedStorageTargetResize, disk):
+        partition = self.get_partition(disk, target.partition_number)
+        part_align = disk.alignment_data().part_align
+        new_size = align_up(target.new_size, part_align)
+        if new_size > partition.size:
+            raise Exception(f'Aligned requested size {new_size} too large')
+        partition.size = new_size
+        partition.resize = True
+        # Calculating where that gap will be can be tricky due to alignment
+        # needs and the possibility that we may be splitting a logical
+        # partition, which needs an extra 1MiB spacer.
+        gap = gaps.after(disk, partition.offset)
+        if gap is None:
+            pgs = gaps.parts_and_gaps(disk)
+            raise Exception(f'gap not found after resize, pgs={pgs}')
+        return gap
+
+    def build_lvm_options(self, password):
+        if password is None:
+            return None
+        else:
+            return {
+                'encrypt': True,
+                'luks_options': {
+                    'password': password,
+                    },
+                }
+
     def guided(self, choice: GuidedChoiceV2):
         self.model.guided_configuration = choice
 
         disk = self.model._one(id=choice.target.disk_id)
-        if isinstance(choice.target, GuidedStorageTargetReformat):
-            mode = 'reformat_disk'
-            target = disk
-        elif isinstance(choice.target, GuidedStorageTargetUseGap):
-            mode = 'use_gap'
-            target = gaps.at_offset(disk, choice.target.gap.offset)
-        elif isinstance(choice.target, GuidedStorageTargetResize):
-            partition = self.get_partition(
-                    disk, choice.target.partition_number)
-            part_align = disk.alignment_data().part_align
-            new_size = align_up(choice.target.new_size, part_align)
-            if new_size > partition.size:
-                raise Exception(f'Aligned requested size {new_size} too large')
-            partition.size = new_size
-            partition.resize = True
-            mode = 'use_gap'
-            # Calculating where that gap will be can be tricky due to alignment
-            # needs and the possibility that we may be splitting a logical
-            # partition, which needs an extra 1MiB spacer.
-            target = gaps.after(disk, partition.offset)
-            if target is None:
-                pgs = gaps.parts_and_gaps(disk)
-                raise Exception(f'gap not found after resize, pgs={pgs}')
-        else:
-            raise Exception(f'Unknown guided target {choice.target}')
+        gap = self.start_guided(choice.target, disk)
+        if DeviceAction.TOGGLE_BOOT in DeviceAction.supported(disk):
+            self.add_boot_disk(disk)
+        # find what's left of the gap after adding boot
+        gap = gap.within()
```

Comment:
Would it make sense to move the responsibility for calling add_boot_disk into start_guided? (maybe a helper to check supported() first I guess...) This thought was triggered by thinking about how this would work for a guided method that would assemble a bunch of disks into a RAID (where you'd want to do something like reformat each disk, call add_boot_disk on each disk, create a partition in the remaining space, create a raid out of those partitions and return the gap for that raid...) but also I think it would let you get rid of the within() games?

Reply:
Yes, I'd love to not have `gap.within()`.

I had an earlier branch that let the boot logic return a gap, I didn't like the resulting code though.  I want to jot this down for future re-exploration.

Reply:
I don't think that's required (although it might be nicer), it would just be changing start_guided_reformat to:

```
    @start_guided.register
    def start_guided_reformat(self, target: GuidedStorageTargetReformat,
                              disk: ModelDisk) -> gaps.Gap:
        """Perform the reformat, and return the resulting gap."""
        self.reformat(disk, wipe='superblock-recursive')
        self.add_boot_disk_if_supported(disk)
        return gaps.largest_gap(disk)
```

I think?